app/models/concerns/cached_counting.rb

```diff
@@ -3,75 +3,160 @@
+      @thread.join
+    end
+  end
 
-    # auto flush if older than this
-    self.autoflush_seconds = 5.minutes
+  def self.enabled?
+    @enabled != false
+  end
 
-    self.last_flush = Time.now.utc
+  def self.enable
+    @enabled = true
   end
 
-  class_methods do
-    def perform_increment!(key, opts = nil)
-      val = DiscourseRedis.ignore_readonly do
-        LUA_INCR_AND_EXPIRE.eval(
-          Discourse.redis.without_namespace,
-          [Discourse.redis.namespace_key(key)],
-          [EXPIRE_CACHE_AFTER]
-        ).to_i
-      end
+  def self.reset
+    clear_queue!
+    clear_flush_to_db_lock!
+  end
 
-      # readonly mode it is going to be nil, skip
-      return if val.nil?
+  def self.ensure_thread!
+    return if !enabled?
 
-      autoflush = (opts && opts[:autoflush]) || self.autoflush
-      if autoflush > 0 && val >= autoflush
-        write_cache!
-        return
+    MUTEX.synchronize do
+      if !@thread&.alive?
+        @thread = nil
       end
+      @thread ||= Thread.new { thread_loop }
+    end
+  end
 
-      if (Time.now.utc - last_flush).to_i > autoflush_seconds
-        write_cache!
+  def self.thread_loop
+    iterations = 0
+    while true
+      break if !enabled?
+
+      sleep SLEEP_SECONDS
+      flush_in_memory
+      if (iterations >= FLUSH_DB_ITERATIONS) || @flush
+        iterations = 0
+        flush_to_db
+        @flush = false
       end
+      iterations += 1
     end
+  end
 
-    def write_cache!(date = nil)
-      raise NotImplementedError
+  def self.flush
+    @flush = true
+    @thread.wakeup
+    while @flush
+      sleep 0.001
     end
+  end
 
-    # this may seem a bit fancy but in so it allows
-    # for concurrent calls without double counting
-    def get_and_reset(key)
-      Discourse.redis.set(key, '0', ex: EXPIRE_CACHE_AFTER, get: true).to_i
+  COUNTER_REDIS_HASH = "CounterCacheHash"
+
+  def self.flush_in_memory
+    counts = nil
+    while QUEUE.length > 0
+      # only 1 consumer, no need to avoid blocking
+      key, klass, db, time = QUEUE.deq
+      _redis_key = "#{klass},#{db},#{time.strftime("%Y%m%d")},#{key}"
+      counts ||= Hash.new(0)
+      counts[_redis_key] += 1
+    end
+
+    if counts
+      counts.each do |redis_key, count|
+        # TODO this whole loop can be done in a single LUA script
+        # concerns:
+        # - Is there a limit of params, will we need to chunk it
+        # - Would this lock up redis for too long, chunk it due to that?
+        Discourse.redis.without_namespace.hincrby(COUNTER_REDIS_HASH, redis_key, count)
+      end
     end
+  end
 
-    def request_id(query_params, retries = 0)
-      id = where(query_params).pluck_first(:id)
-      id ||= create!(query_params.merge(count: 0)).id
-    rescue # primary key violation
-      if retries == 0
-        request_id(query_params, 1)
-      else
-        raise
+  DB_FLUSH_COOLDOWN_SECONDS = 60
+  DB_COOLDOWN_KEY = "cached_counting_cooldown"
+
+  def self.flush_to_db
```

Comment:
Similar to above, I think we'll lose data under Postgres readonly. Do we care? 🤔